README.md
pyprimesieve
============

Many primes, very fast. Uses [primesieve][0].

primesieve, one of the fastest (if not the fastest) prime sieve implementaions available, is actively maintained by
Kim Walisch.

It uses a segmented sieve of Eratosthenes with wheel factorization for a complexity of `O(nloglogn)` operations.


Performance
-----------

Regarding primesieve for C++:

> primesieve generates the first 50,847,534 primes up to 10^9 in just 0.4 seconds on a single core of an Intel Core
> i7-920 2.66GHz, this is about 50 times faster than an ordinary C/C++ sieve of Eratosthenes implementation and about
> 10,000 times faster than trial-division. primesieve outperforms [Kim's] older [ecprime][1] (fastest from 2002 to 2010) by
> about 30 percent and also substantially outperforms [primegen][2] the fastest sieve of Atkin implementation on the
> web.

For comparison, on an Intel Core i7 2GHz, `pyprimesieve` populates an entire Python list of the first
50,847,534 primes in 1.40 seconds. It's expected that a Python implementation would be slower than C++ but,
surprisingly, by only one second.

`pyprimesieve` outperforms all of the fastest prime sieving implementations for Python.

    Time (ms) to generate the all primes below one million and iterate over them in Python:

    pyprimesieve         2.79903411865
    primesfrom2to        13.1568908691
    primesfrom3to        13.5800838470
    ambi_sieve           16.1600112915
    rwh_primes2          38.7749671936
    rwh_primes1          48.5658645630
    rwh_primes           52.0040988922
    sieve_wheel_30       59.3869686127
    sieveOfEratosthenes  59.4990253448
    ambi_sieve_plain     161.740064621
    sieveOfAtkin         232.724905014
    sundaram3            251.194953918

It can be seen here that `pyprimesieve` is *4.7 times faster* than the fastest Python alternative using `Numpy` and
*13.85 times faster* than the fastest pure Python sieve.

All benchmark scripts and algorithms are available for reproduction. Prime sieve algorithm implementations were taken
from [this discussion on SO][3].

Functions
---------

**primes(n)**: List of prime numbers up to `n`.

**primes(start, n)**: List of prime numbers from `start` up to `n`.

**primes_sum(n)**: The summation of prime numbers up to `n`. The optimal number of threads will be determined for the
given number and system.

**primes_sum(start, n)**: The summation of prime numbers from `start` up to `n`. The optimal number of threads will be
determined for the given numbers and system.

**primes_nth(n)**: The nth prime number.

**factorize(n)**: List of tuples in the form of (prime, power) for the prime factorization of `n`.


Installation
------------

    python setup.py install

NOTE: Because of the need to use OpenMP to compile the parallelized version of summation, g++ is specified in
environment variables of setup to avoid distutils choosing a compiler that does not have support for OpenMP. If you
don't have g++, you will need to change that in `setup.py`.

Like any C/C++ extension, you need to have the development package of Python (able to include `Python.h`) in order to
compile. On Ubuntu you can simply run,

    sudo apt-get install python-dev

After installation, you can make sure everything is working by running the following inside the project root folder,

    python tests


License
-------

"Modified BSD License". See LICENSE for details. Copyright Jared Suttles, 2013.

[0]: http://code.google.com/p/primesieve/
[1]: http://primzahlen.de/referenten/Kim_Walisch/index2.htm
[2]: http://cr.yp.to/primegen.html
[3]: http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python


CHANGES
Changes
=======

0.1.3 (2013-08-05)
------------------

- Fixed compilation warnings
- extern "C"
- Added more information regarding installation to README; updated docstring


0.1.2 (2013-07-12)
------------------

- primes_nth uses callback class instead of creating any PyObjects in order to call primes
    - Caused a memory leak and was unnecessary


0.1.1 (2013-07-12)
------------------

- Distutils setup.py: got multithreaded summation compiling with OpenMP
    - Removed all code to manually compile executable
    - CC and CXX environment variables set to avoid clang (and other non-openmp compilers) use
- Replaced executable code with parallelized function in the actual extension
- Rewrote primes_sum and primes_nth Python functions in C++ and removed Python script completely


0.1.0 (2013-07-06)
------------------

- pyprimesieve initial


tests/__main__.py
#!/usr/bin/env python
import unittest
from test_primes import TestPrimes
from test_primes_sum import TestSumPrimes
from test_factorize import TestFactorize
from test_primes_nth import TestNthPrimes

suite = unittest.TestSuite()
suite.addTests([unittest.TestLoader().loadTestsFromTestCase(TestPrimes),
                unittest.TestLoader().loadTestsFromTestCase(TestSumPrimes),
                unittest.TestLoader().loadTestsFromTestCase(TestFactorize),
                unittest.TestLoader().loadTestsFromTestCase(TestNthPrimes)])

#runner = unittest.TextTestRunner()
#runner.run(suite)
unittest.TextTestRunner().run(suite)


tests/test_primes.py
#!/usr/bin/env python
import unittest
import pyprimesieve
from itertools import izip_longest, dropwhile
try:
    from numpy import ones, bool, r_, nonzero, ndarray, allclose
    HAS_NUMPY = True
except ImportError:
    HAS_NUMPY = False


def _primes_numpy(n):
    """
    Input n>=6, Returns a array of primes, 2 <= p < n
    http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python/3035188#3035188
    """
    sieve = ones(n/3 + (n%6==2), dtype=bool)
    for i in xrange(1,int(n**0.5)/3+1):
        if sieve[i]:
            k=3*i+1|1
            sieve[       k*k/3     ::2*k] = False
            sieve[k*(k-2*(i&1)+4)/3::2*k] = False
    return r_[2,3,((3*nonzero(sieve)[0][1:]+1)|1)]


def _primes(n):
    """
    Input n>=6, Returns a list of primes, 2 <= p < n
    http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python/3035188#3035188
    """
    n, correction = n-n%6+6, 2-(n%6>1)
    sieve = [True] * (n/3)
    for i in xrange(1,int(n**0.5)/3+1):
      if sieve[i]:
        k=3*i+1|1
        sieve[      k*k/3      ::2*k] = [False] * ((n/6-k*k/6-1)/k+1)
        sieve[k*(k-2*(i&1)+4)/3::2*k] = [False] * ((n/6-k*(k-2*(i&1)+4)/6-1)/k+1)
    return [2,3] + [3*i+1|1 for i in xrange(1,n/3-correction) if sieve[i]]


class TestPrimes(unittest.TestCase):
    def test_1(self):
        self.assertEqual(pyprimesieve.primes(1), [])

    def test_2(self):
        self.assertEqual(pyprimesieve.primes(2), [])

    def test_3(self):
        self.assertEqual(pyprimesieve.primes(3), [2])

    def test_4(self):
        self.assertEqual(pyprimesieve.primes(4), [2, 3])

    def test_prime_count_1(self):
        self.assertEqual(len(pyprimesieve.primes(10**7)), 664579)

    def test_prime_count_2(self):
        self.assertEqual(len(pyprimesieve.primes(10**8)), 5761455)

    #def test_prime_count_3(self):
    #    self.assertEqual(len(pyprimesieve.primes(10**9)), 50847534)

    def test_emptylist_1(self):
        self.assertEqual(pyprimesieve.primes(0), [])

    def test_emptylist_2(self):
        self.assertEqual(pyprimesieve.primes(-1), [])

    def test_ranges_1(self):
        # start > n
        self.assertEqual(pyprimesieve.primes(411, 42), [])

    def test_ranges_2(self):
        # start < 0 - should be uneffected as if it were 2
        self.assertTrue(sequences_equal(pyprimesieve.primes(-100, 42), primes(42)))

    def test_ranges_3(self):
        # arbitrary point to start
        self.assertTrue(sequences_equal(pyprimesieve.primes(1412, 85747),
                                        dropwhile(lambda n: n < 1412, primes(85747))))

    def test_ranges_4(self):
        # arbitrary point to start
        self.assertTrue(sequences_equal(pyprimesieve.primes(74651, 975145),
                                        dropwhile(lambda n: n < 74651, primes(975145))))


def sequences_equal(lst1, lst2):
    return all(a == b for a, b in izip_longest(lst1, lst2))

primes = _primes_numpy if HAS_NUMPY else _primes

for i, n in enumerate(xrange(100, 10000, 100)):  # create sequence comparison tests for sieves of size n in range
    test = lambda self: self.assertTrue(sequences_equal(pyprimesieve.primes(n), primes(n)))
    setattr(TestPrimes, 'test_' + str(i), test)


if __name__ == "__main__":
    unittest.main()


tests/test_factorize.py
#!/usr/bin/env python
import unittest
import pyprimesieve
from test_primes import sequences_equal
import random


def factorize(n):
    """
    List of the prime factorization of the given number represented by tuples of the prime and power.
    """
    prime_factors = []
    if n < 2:
        return prime_factors
    i = 0
    while not n & 1:
        n /= 2
        i += 1
    if i > 0:
        prime_factors.append((2, i))
    for p in xrange(3, n, 2):  # use n here but won't ever be > int(sqrt(n))+1
        if p*p > n:
            break
        i = 0
        while n % p == 0:
            n /= p
            i += 1
        if i > 0:
            prime_factors.append((p, i))
    if n > 1:
        prime_factors.append((n, 1))
    return prime_factors


class TestFactorize(unittest.TestCase):
    def test_negative_1(self):
        self.assertEqual(pyprimesieve.factorize(-1), [])

    def test_negative_2(self):
        self.assertEqual(pyprimesieve.factorize(-48485), [])

l = range(10**6)
random.shuffle(l)
l = l[:10**4]
for i, n in enumerate(l):
    test = lambda self: self.assertTrue(sequences_equal(pyprimesieve.factorize(n), factorize(n)))
    setattr(TestFactorize, 'test_' + str(i), test)

if __name__ == "__main__":
    unittest.main()


pyprimesieve/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# pyprimesieve: Many primes, very fast. Uses primesieve.
# Copyright: (c) 2013, Jared Suttles. All rights reserved.
# License: BSD, see LICENSE for details.
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

"""
pyprimesieve: Many primes, very fast. Uses primesieve.
- - - - - - - - - - - - - - - - - - - - - - - - - - -

primes(n): List of prime numbers up to `n`.

primes(start, n): List of prime numbers from `start` up to `n`.

primes_sum(n [, threads]): The summation of prime numbers up to `n`. If `threads` is given, that many threads will
be created. If not, the optimal number of threads will be determined.

primes_sum(start, n [, threads]): The summation of prime numbers from `start` up to `n`. If `threads` is given,
that many threads will be created. If not, the optimal number of threads will be determined.

primes_nth(n): The nth prime number.

factorize(n): List of tuples in the form of (*prime*, *power*) for the prime factorization of `n`.


Copyright: (c) 2013, Jared Suttles. All rights reserved.
License: BSD, see LICENSE for details.

"""

from subprocess import Popen
from multiprocessing import cpu_count
from os import path
from math import log
try:
    from ._pyprimesieve import primes, factorize
except ImportError:
    pass

__all__ = __dir__ = ['primes', 'primes_sum', 'primes_nth', 'factorize']
__title__ = 'pyprimesieve'
__version__ = '0.1.0'
__author__ = 'Jared Suttles'
__license__ = 'BSD License'
__copyright__ = 'Copyright 2013 Jared Suttles'


def primes_sum(*args, **kwargs):
    """
    The summation of prime numbers from `start` up to `n`. If `threads` is given, that many threads will be created.
    If not, the optimal number of threads will be determined.
    """
    if not args:
        raise TypeError('primes_sum expected at least 1 arguments, got 0')
    if len(args) > 2:
        raise TypeError('function takes at most 2 arguments ({} given)'.format(len(args)))
    try:
        start, n = args[0], args[1]
    except IndexError:
        start, n = 0, args[0]
    try:
        start, n = int(start), int(n)
    except ValueError:
        raise TypeError('an integer is required')
    if start > n or n < 3:
        return 0
    if start < 2:
        start = 2
    if n < primes_sum.threshold:
        return sum(primes(start, n))
    threads = kwargs.get('threads')
    if threads is None:
        try:
            threads = cpu_count()
        except NotImplementedError:
            threads = 1
    elif threads < 1:
        threads = 1
    pyprimes_primes_sum = path.join(path.dirname(path.realpath(__file__)), 'pyprimesieve_sum')
    try:
        p = Popen([pyprimes_primes_sum, str(n), str(start), str(threads)], stdout=-1)
        return int(p.stdout.read())
    except OSError:
        return sum(primes(start, n))
primes_sum.threshold = 2500000


def primes_nth(n):
    """
    The Nth prime number.
    """
    if n < 1:
        raise ValueError('must provide a positive number')
    elif n < 6:
        return [2, 3, 5, 7, 11][n-1]
    return primes(int(n*log(n*log(n))))[n-1]


bench/bench_primes.py
#!/usr/bin/env python
from timeit import Timer
from math import sqrt, ceil
import numpy as np
from pyprimesieve import primes as pyprimesieve


################################################################################
# Prime algorithms
################################################################################
def rwh_primes(n):
    # http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python/3035188#3035188
    """ Returns  a list of primes < n """
    sieve = [True] * n
    for i in xrange(3,int(n**0.5)+1,2):
        if sieve[i]:
            sieve[i*i::2*i]=[False]*((n-i*i-1)/(2*i)+1)
    return [2] + [i for i in xrange(3,n,2) if sieve[i]]

def rwh_primes1(n):
    # http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python/3035188#3035188
    """ Returns  a list of primes < n """
    sieve = [True] * (n/2)
    for i in xrange(3,int(n**0.5)+1,2):
        if sieve[i/2]:
            sieve[i*i/2::i] = [False] * ((n-i*i-1)/(2*i)+1)
    return [2] + [2*i+1 for i in xrange(1,n/2) if sieve[i]]

def rwh_primes2(n):
    # http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python/3035188#3035188
    """ Input n>=6, Returns a list of primes, 2 <= p < n """
    correction = (n%6>1)
    n = {0:n,1:n-1,2:n+4,3:n+3,4:n+2,5:n+1}[n%6]
    sieve = [True] * (n/3)
    sieve[0] = False
    for i in xrange(int(n**0.5)/3+1):
      if sieve[i]:
        k=3*i+1|1
        sieve[      ((k*k)/3)      ::2*k]=[False]*((n/6-(k*k)/6-1)/k+1)
        sieve[(k*k+4*k-2*k*(i&1))/3::2*k]=[False]*((n/6-(k*k+4*k-2*k*(i&1))/6-1)/k+1)
    return [2,3] + [3*i+1|1 for i in xrange(1,n/3-correction) if sieve[i]]

def sieve_wheel_30(N):
    # http://zerovolt.com/?p=88
    '''
    Returns a list of primes <= N using wheel criterion 2*3*5 = 30
    Copyright 2009 by zerovolt.com
    This code is free for non-commercial purposes, in which case you can just leave this comment as a credit for my
    work.
    If you need this code for commercial purposes, please contact me by sending an email to: info [at] zerovolt [dot]
    com.
    '''
    __smallp = ( 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59,
    61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139,
    149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227,
    229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311,
    313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401,
    409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491,
    499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599,
    601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683,
    691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797,
    809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887,
    907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997)

    wheel = (2, 3, 5)
    const = 30
    if N < 2:
        return []
    if N <= const:
        pos = 0
        while __smallp[pos] <= N:
            pos += 1
        return list(__smallp[:pos])
    # make the offsets list
    offsets = (7, 11, 13, 17, 19, 23, 29, 1)
    # prepare the list
    p = [2, 3, 5]
    dim = 2 + N // const
    tk1  = [True] * dim
    tk7  = [True] * dim
    tk11 = [True] * dim
    tk13 = [True] * dim
    tk17 = [True] * dim
    tk19 = [True] * dim
    tk23 = [True] * dim
    tk29 = [True] * dim
    tk1[0] = False
    # help dictionary d
    # d[a , b] = c  ==> if I want to find the smallest useful multiple of (30*pos)+a
    # on tkc, then I need the index given by the product of [(30*pos)+a][(30*pos)+b]
    # in general. If b < a, I need [(30*pos)+a][(30*(pos+1))+b]
    d = {}
    for x in offsets:
        for y in offsets:
            res = (x*y) % const
            if res in offsets:
                d[(x, res)] = y
    # another help dictionary: gives tkx calling tmptk[x]
    tmptk = {1:tk1, 7:tk7, 11:tk11, 13:tk13, 17:tk17, 19:tk19, 23:tk23, 29:tk29}
    pos, prime, lastadded, stop = 0, 0, 0, int(ceil(sqrt(N)))
    # inner functions definition
    def del_mult(tk, start, step):
        for k in xrange(start, len(tk), step):
            tk[k] = False
    # end of inner functions definition
    cpos = const * pos
    while prime < stop:
        # 30k + 7
        if tk7[pos]:
            prime = cpos + 7
            p.append(prime)
            lastadded = 7
            for off in offsets:
                tmp = d[(7, off)]
                start = (pos + prime) if off == 7 else (prime * (const * (pos + 1 if tmp < 7 else 0) + tmp) )//const
                del_mult(tmptk[off], start, prime)
        # 30k + 11
        if tk11[pos]:
            prime = cpos + 11
            p.append(prime)
            lastadded = 11
            for off in offsets:
                tmp = d[(11, off)]
                start = (pos + prime) if off == 11 else (prime * (const * (pos + 1 if tmp < 11 else 0) + tmp) )//const
                del_mult(tmptk[off], start, prime)
        # 30k + 13
        if tk13[pos]:
            prime = cpos + 13
            p.append(prime)
            lastadded = 13
            for off in offsets:
                tmp = d[(13, off)]
                start = (pos + prime) if off == 13 else (prime * (const * (pos + 1 if tmp < 13 else 0) + tmp) )//const
                del_mult(tmptk[off], start, prime)
        # 30k + 17
        if tk17[pos]:
            prime = cpos + 17
            p.append(prime)
            lastadded = 17
            for off in offsets:
                tmp = d[(17, off)]
                start = (pos + prime) if off == 17 else (prime * (const * (pos + 1 if tmp < 17 else 0) + tmp) )//const
                del_mult(tmptk[off], start, prime)
        # 30k + 19
        if tk19[pos]:
            prime = cpos + 19
            p.append(prime)
            lastadded = 19
            for off in offsets:
                tmp = d[(19, off)]
                start = (pos + prime) if off == 19 else (prime * (const * (pos + 1 if tmp < 19 else 0) + tmp) )//const
                del_mult(tmptk[off], start, prime)
        # 30k + 23
        if tk23[pos]:
            prime = cpos + 23
            p.append(prime)
            lastadded = 23
            for off in offsets:
                tmp = d[(23, off)]
                start = (pos + prime) if off == 23 else (prime * (const * (pos + 1 if tmp < 23 else 0) + tmp) )//const
                del_mult(tmptk[off], start, prime)
        # 30k + 29
        if tk29[pos]:
            prime = cpos + 29
            p.append(prime)
            lastadded = 29
            for off in offsets:
                tmp = d[(29, off)]
                start = (pos + prime) if off == 29 else (prime * (const * (pos + 1 if tmp < 29 else 0) + tmp) )//const
                del_mult(tmptk[off], start, prime)
        # now we go back to top tk1, so we need to increase pos by 1
        pos += 1
        cpos = const * pos
        # 30k + 1
        if tk1[pos]:
            prime = cpos + 1
            p.append(prime)
            lastadded = 1
            for off in offsets:
                tmp = d[(1, off)]
                start = (pos + prime) if off == 1 else (prime * (const * pos + tmp) )//const
                del_mult(tmptk[off], start, prime)
    # time to add remaining primes
    # if lastadded == 1, remove last element and start adding them from tk1
    # this way we don't need an "if" within the last while
    if lastadded == 1:
        p.pop()
    # now complete for every other possible prime
    while pos < len(tk1):
        cpos = const * pos
        if tk1[pos]: p.append(cpos + 1)
        if tk7[pos]: p.append(cpos + 7)
        if tk11[pos]: p.append(cpos + 11)
        if tk13[pos]: p.append(cpos + 13)
        if tk17[pos]: p.append(cpos + 17)
        if tk19[pos]: p.append(cpos + 19)
        if tk23[pos]: p.append(cpos + 23)
        if tk29[pos]: p.append(cpos + 29)
        pos += 1
    # remove exceeding if present
    pos = len(p) - 1
    while p[pos] > N:
        pos -= 1
    if pos < len(p) - 1:
        del p[pos+1:]
    # return p list
    return p

def sieveOfEratosthenes(n):
    """sieveOfEratosthenes(n): return the list of the primes < n."""
    # Code from: <dickinsm@gmail.com>, Nov 30 2006
    # http://groups.google.com/group/comp.lang.python/msg/f1f10ced88c68c2d
    if n <= 2:
        return []
    sieve = range(3, n, 2)
    top = len(sieve)
    for si in sieve:
        if si:
            bottom = (si*si - 3) // 2
            if bottom >= top:
                break
            sieve[bottom::si] = [0] * -((bottom - top) // si)
    return [2] + [el for el in sieve if el]

def sieveOfAtkin(end):
    """sieveOfAtkin(end): return a list of all the prime numbers <end
    using the Sieve of Atkin."""
    # Code by Steve Krenzel, <Sgk284@gmail.com>, improved
    # Code: http://krenzel.info/?p=83
    # Info: http://en.wikipedia.org/wiki/Sieve_of_Atkin
    assert end > 0
    lng = ((end-1) // 2)
    sieve = [False] * (lng + 1)

    x_max, x2, xd = int(sqrt((end-1)/4.0)), 0, 4
    for xd in xrange(4, 8*x_max + 2, 8):
        x2 += xd
        y_max = int(sqrt(end-x2))
        n, n_diff = x2 + y_max*y_max, (y_max << 1) - 1
        if not (n & 1):
            n -= n_diff
            n_diff -= 2
        for d in xrange((n_diff - 1) << 1, -1, -8):
            m = n % 12
            if m == 1 or m == 5:
                m = n >> 1
                sieve[m] = not sieve[m]
            n -= d

    x_max, x2, xd = int(sqrt((end-1) / 3.0)), 0, 3
    for xd in xrange(3, 6 * x_max + 2, 6):
        x2 += xd
        y_max = int(sqrt(end-x2))
        n, n_diff = x2 + y_max*y_max, (y_max << 1) - 1
        if not(n & 1):
            n -= n_diff
            n_diff -= 2
        for d in xrange((n_diff - 1) << 1, -1, -8):
            if n % 12 == 7:
                m = n >> 1
                sieve[m] = not sieve[m]
            n -= d

    x_max, y_min, x2, xd = int((2 + sqrt(4-8*(1-end)))/4), -1, 0, 3
    for x in xrange(1, x_max + 1):
        x2 += xd
        xd += 6
        if x2 >= end: y_min = (((int(ceil(sqrt(x2 - end))) - 1) << 1) - 2) << 1
        n, n_diff = ((x*x + x) << 1) - 1, (((x-1) << 1) - 2) << 1
        for d in xrange(n_diff, y_min, -8):
            if n % 12 == 11:
                m = n >> 1
                sieve[m] = not sieve[m]
            n += d

    primes = [2, 3]
    if end <= 3:
        return primes[:max(0,end-2)]

    for n in xrange(5 >> 1, (int(sqrt(end))+1) >> 1):
        if sieve[n]:
            primes.append((n << 1) + 1)
            aux = (n << 1) + 1
            aux *= aux
            for k in xrange(aux, end, 2 * aux):
                sieve[k >> 1] = False

    s  = int(sqrt(end)) + 1
    if s  % 2 == 0:
        s += 1
    primes.extend([i for i in xrange(s, end, 2) if sieve[i >> 1]])

    return primes

def ambi_sieve_plain(n):
    s = range(3, n, 2)
    for m in xrange(3, int(n**0.5)+1, 2):
        if s[(m-3)/2]:
            for t in xrange((m*m-3)/2,(n>>1)-1,m):
                s[t]=0
    return [2]+[t for t in s if t>0]

def sundaram3(max_n):
    # http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python/2073279#2073279
    numbers = range(3, max_n+1, 2)
    half = (max_n)//2
    initial = 4

    for step in xrange(3, max_n+1, 2):
        for i in xrange(initial, half, step):
            numbers[i-1] = 0
        initial += 2*(step+1)

        if initial > half:
            return [2] + filter(None, numbers)


################################################################################
# Using Numpy
################################################################################
def ambi_sieve(n):
    # http://tommih.blogspot.com/2009/04/fast-prime-number-generator.html
    s = np.arange(3, n, 2)
    for m in xrange(3, int(n ** 0.5)+1, 2):
        if s[(m-3)/2]:
            s[(m*m-3)/2::m]=0
    return np.r_[2, s[s>0]]

def primesfrom3to(n):
    # http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python/3035188#3035188
    """ Returns a array of primes, p < n """
    assert n>=2
    sieve = np.ones(n/2, dtype=np.bool)
    for i in xrange(3,int(n**0.5)+1,2):
        if sieve[i/2]:
            sieve[i*i/2::i] = False
    return np.r_[2, 2*np.nonzero(sieve)[0][1::]+1]

def primesfrom2to(n):
    # http://stackoverflow.com/questions/2068372/fastest-way-to-list-all-primes-below-n-in-python/3035188#3035188
    """ Input n>=6, Returns a array of primes, 2 <= p < n """
    sieve = np.ones(n/3 + (n%6==2), dtype=np.bool)
    sieve[0] = False
    for i in xrange(int(n**0.5)/3+1):
        if sieve[i]:
            k=3*i+1|1
            sieve[      ((k*k)/3)      ::2*k] = False
            sieve[(k*k+4*k-2*k*(i&1))/3::2*k] = False
    return np.r_[2,3,((3*np.nonzero(sieve)[0]+1)|1)]


################################################################################
# Performance testing
################################################################################
def iterate_primes(primes_function):  # iteration speed matters as well
    def find_primes_and_iterate(n):
        for _ in primes_function(n):
            pass
    return find_primes_and_iterate


def benchmark_primes(n, functions, bestof=6):
    algorithm_times = []
    for function_name in functions:
        timer = min(Timer(stmt='fn = iterate_primes({}); fn({})'.format(function_name, n),
                          setup='from __main__ import {}, iterate_primes'.format(function_name)).repeat(repeat=bestof,
                                                                                                        number=1))
        timer *= 10**3
        algorithm_times.append((function_name, timer))
    return sorted(algorithm_times, key=lambda x: x[1])


FUNCTIONS = [
    'pyprimesieve', 'rwh_primes', 'rwh_primes1', 'rwh_primes2', 'sieve_wheel_30', 'sieveOfEratosthenes',
    'sieveOfAtkin', 'ambi_sieve_plain', 'sundaram3', 'ambi_sieve', 'primesfrom3to', 'primesfrom2to',
]


if __name__ == "__main__":
    for n in xrange(6, 7):
        print '\n\n  Time to generate 10**{} primes and iterate over them in Python'.format(n)
        print '_' * 80

        n = 10**n
        for name, time in benchmark_primes(n, FUNCTIONS):
            print '{} -> {}'.format(name, time)
